streams/src/main/java/org/apache/kafka/streams/processor/assignment/assignors/StickyTaskAssignor.java

```diff
@@ -0,0 +1,478 @@
+    }
+
+    private void optimizeStandby(final ApplicationState applicationState, final AssignmentState assignmentState) {
+        if (applicationState.assignmentConfigs().numStandbyReplicas() <= 0) {
+            return;
+        }
+
+        if (!TaskAssignmentUtils.hasValidRackInformation(applicationState)) {
+            return;
+        }
+
+        final Map<ProcessId, KafkaStreamsAssignment> currentAssignments = assignmentState.buildKafkaStreamsAssignments();
+        final Map<ProcessId, KafkaStreamsAssignment> optimizedAssignments = TaskAssignmentUtils.optimizeRackAwareStandbyTasks(
+            applicationState, currentAssignments);
+        assignmentState.processOptimizedAssignments(optimizedAssignments);
+    }
+
+    private static void assignActive(final ApplicationState applicationState,
+                                     final Collection<KafkaStreamsState> clients,
+                                     final AssignmentState assignmentState,
+                                     final boolean mustPreserveActiveTaskAssignment) {
+        final int totalCapacity = computeStreamThreadCount(clients);
+        if (totalCapacity == 0) {
+            throw new IllegalStateException("`totalCapacity` should never be zero.");
+        }
+
+        final Set<TaskId> allTaskIds = applicationState.allTasks().stream()
+            .map(TaskInfo::id).collect(Collectors.toSet());
+        final int taskCount = allTaskIds.size();
+        final int activeTasksPerThread = taskCount / totalCapacity;
+        final Set<TaskId> unassigned = new HashSet<>(allTaskIds);
+
+        // first try and re-assign existing active tasks to clients that previously had
+        // the same active task
+        for (final TaskId taskId : assignmentState.previousActiveAssignment.keySet()) {
+            final ProcessId previousClientForTask = assignmentState.previousActiveAssignment.get(taskId);
+
+            // TODO: Remove this check, why is it even necessary?
+            if (allTaskIds.contains(taskId)) {
+                if (mustPreserveActiveTaskAssignment || assignmentState.hasRoomForActiveTask(previousClientForTask, activeTasksPerThread)) {
+                    assignmentState.finalizeAssignment(taskId, previousClientForTask, AssignedTask.Type.ACTIVE);
+                    unassigned.remove(taskId);
+                }
+            }
+        }
+
+        // try and assign any remaining unassigned tasks to clients that previously
+        // have seen the task.
+        for (final Iterator<TaskId> iterator = unassigned.iterator(); iterator.hasNext(); ) {
+            final TaskId taskId = iterator.next();
+            final Set<ProcessId> previousClientsForStandbyTask = assignmentState.previousStandbyAssignment.get(taskId);
+            if (previousClientsForStandbyTask == null || previousClientsForStandbyTask.isEmpty()) {
+                continue;
+            }
```

Comment:
nit: can we try and avoid this style of conditional early-exit, at least inside loops? I found the original version of this code a little clearer, it's easier to follow the logic if it's all in one block (and the previousClientsForStandbyTask.isEmpty() check does nothing since it won't enter the for loop anyways if the set is empty)